### PR TITLE
ENH: covariance estimator in onedal4py/sklearnex

### PR DESCRIPTION
**Pull request by olegkkruglov**
Changes proposed in this pull request:
- Covariance algorithm is moved from primitives to algorithms
- sklearn EmpiricalCovariance is patched now (only if assume_centered == False and store_precision == False)

### CODE REVIEW

**Comment by samir-nasibli** [comment]
@olegkkruglov please add tests for onedal4py module for different devices offloading

File: onedal/covariance/covariance.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        result = module.compute(policy, params, to_table(X))
+        self.covariance_ = from_table(result.cov_matrix)
+        self.location_ = from_table(result.means).ravel()
+
+        return self
+
+
+class Covariance(BaseCovariance):
+    def __init__(self):
+        pass
```
Comment (samir-nasibli) [inline onedal/covariance/covariance.py:59]:
```suggestion
    def __init__(self, method="by_default"):
        super().__init__(method)
```

File: onedal/covariance/covariance.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from daal4py.sklearn._utils import get_dtype, make2d
+from onedal import _backend
+
+from ..common._policy import _get_policy
+from ..datatypes import _convert_to_supported, from_table, to_table
+
+
+class BaseCovariance:
+    def __init__(self):
+        pass
```
Comment (samir-nasibli) [inline onedal/covariance/covariance.py:29]:
```suggestion
    def __init__(self, method):
        self.method= method
```

File: onedal/covariance/covariance.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    def _get_policy(self, queue, *data):
+        return _get_policy(queue, *data)
+
+    def _get_onedal_params(self, dtype=np.float32):
+        return {
+            "fptype": "float" if dtype == np.float32 else "double",
+            "method": "dense",
```
Comment (samir-nasibli) [inline onedal/covariance/covariance.py:37]:
```suggestion
            "method": self.method,
```

**Comment by samir-nasibli** [comment]
@olegkkruglov please also update https://github.com/intel/scikit-learn-intelex/blob/37fa33fdc6a2c37e7ecb946742c7a91ee781d67c/onedal/__init__.py#L46
```suggestion
 __all__ = ["covariance", "decomposition", "ensemble", "neighbors", "primitives", "svm"] 
```
It will require some updates in your code also.

File: onedal/covariance/covariance.py (hunk 2 above)
Comment (olegkkruglov) [reply onedal/covariance/covariance.py:29]: I think these changes are redundant since we only have one available method for Covariance at this moment.

Comment (samir-nasibli) [reply onedal/covariance/covariance.py:29]: We have this on our OneDAL API, it worth have it for onedal4py and sklearnex ifaces as well.
For some of our OneDAL estimators we sometimes have params with one able setting, so for consistency it worth have here as well.

**Comment by samir-nasibli** [comment]
@olegkkruglov please address https://github.com/intel/scikit-learn-intelex/pull/1569#discussion_r1385120819 and then run also intel CI

File: onedal/covariance/covariance.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        params = self._get_onedal_params(dtype)
+        result = module.compute(policy, params, to_table(X))
+        self.covariance_ = from_table(result.cov_matrix)
+        self.location_ = from_table(result.means).ravel()
+
+        return self
+
+
+class Covariance(BaseCovariance):
+    def __init__(self, method="dense"):
```
Comment (samir-nasibli) [inline onedal/covariance/covariance.py:58]: It worth to have some minimal doc strings here.
As a reference use this https://github.com/scikit-learn/scikit-learn/blob/093e0cf14aff026cca6097e8c42f83b735d26358/sklearn/covariance/_empirical_covariance.py#L109

**Comment by samir-nasibli** [comment]
Thank you @olegkkruglov ! LGTM. let's wait CI checks now

File: onedal/covariance/covariance.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        self.covariance_ = from_table(result.cov_matrix)
+        self.location_ = from_table(result.means).ravel()
+
+        return self
+
+
+class Covariance(BaseCovariance):
+    """Covariance estimator.
+
+    Computes sample covariance matrix.
```
Comment (aepanchi) [inline onedal/covariance/covariance.py:60]:
```suggestion
    Computes sample covariance matrix.
```

File: onedal/covariance/covariance.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    Parameters
+    ----------
+    method : string, default="dense"
+        Specifies computation method. Available methods: "dense".
+
+    Attributes
+    ----------
+    location_ : ndarray of shape (n_features,)
+        Estimated location, i.e. the estimated mean.
```
Comment (aepanchi) [inline onedal/covariance/covariance.py:70]:
```suggestion
        Estimated location, i.e., the estimated mean.
```

File: onedal/covariance/covariance.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    def __init__(self, method="dense"):
+        super().__init__(method)
+
+    def fit(self, X, queue=None):
+        """Fit the sample covariance matrix of X.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training data, where `n_samples` is the number of samples and
```
Comment (aepanchi) [inline onedal/covariance/covariance.py:85]:
```suggestion
            Training data, where `n_samples` is the number of samples, and
```

File: onedal/covariance/covariance.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            `n_features` is the number of features.
+
+        queue : dpctl.SyclQueue
+            If not None then use this queue for computations.
```
Comment (aepanchi) [inline onedal/covariance/covariance.py:89]:
```suggestion
            If not None, use this queue for computations.
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
Changes for module already existing in default sklearn requires broader discussion.
Particularly, how is this estimator connected to `sklearn.covariance.EmpiricalCovariance` and `sklearn.covariance.empirical_covariance`?

**Comment by olegkkruglov** [comment]
It is different algorithms. EmpiricalCovariance from sklearn computes maximum likelihood estimation of covariance which is not implemented in oneDAL as far as I know. This algorithm is sample covariance estimation which is different

File: onedal/covariance/covariance.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    def _get_onedal_params(self, dtype=np.float32):
+        return {
+            "fptype": "float" if dtype == np.float32 else "double",
+            "method": self.method,
+        }
+
+    def _fit(self, X, module, queue):
+        policy = self._get_policy(queue, X)
+        X = check_array(X, dtype=[np.float64, np.float32])
```
Comment (samir-nasibli) [inline onedal/covariance/covariance.py:42]:
```suggestion
        X = check_array(X, dtype=[np.float64, np.float32], ensure_2d=False, force_all_finite=False)
```

**Comment by olegkkruglov** [comment]
> It is different algorithms. EmpiricalCovariance from sklearn computes maximum likelihood estimation of covariance which is not implemented in oneDAL as far as I know. This algorithm is sample covariance estimation which is different. What changes are necessary in this case?

Sorry, I was wrong. I have checked more deep and realized that these algorithms are almost the same and result of one of them can be obtained from result of the other one by multiplying by scalar. Probably it means that we should patch sklearn.EmpiricalCovariance. I will update the PR soon.

File: sklearnex/covariance/tests/test_covariance.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+def test_sklearnex_import_covariance():
+    from sklearnex.covariance import EmpiricalCovariance
+
+    X = np.array([[0, 1], [0, 1]])
+    result = EmpiricalCovariance().fit(X)
+    expected_covariance = np.array([[0, 0], [0, 0]])
+    expected_means = np.array([0, 1])
+
+    assert_allclose(expected_covariance, result.covariance_)
+    assert_allclose(expected_means, result.location_)
```
Comment (samir-nasibli) [inline sklearnex/covariance/tests/test_covariance.py:31]: It is better to check it for different dataframes and devices.
Like here: https://github.com/intel/scikit-learn-intelex/blob/master/sklearnex/preview/decomposition/tests/test_preview_pca.py#L29:L38

**Review by samir-nasibli — COMMENTED** [review]
Thank you @olegkkruglov
Generally looks good to me.
Just address some comments and let's wait CI checks.

This sklearnex Covariance should be compared with stock sklearn impl via [sklearn bench tool](https://github.com/IntelPython/scikit-learn_bench) for considering it on main branch.

Also OneDAL Covariance supports SPMD intreface, that should be also added and the work should be done in the separate PR.

**Comment by samir-nasibli** [comment]
@olegkkruglov generally looks goods to me.
Please just address some test fails from internal CI
```
FAILED tests/test_common.py::test_estimators[EmpiricalCovariance()-check_fit_score_takes_y]
FAILED tests/test_common.py::test_estimators[EmpiricalCovariance()-check_pipeline_consistency]
FAILED tests/test_common.py::test_estimators[EmpiricalCovariance()-check_n_features_in]
```

File: sklearnex/dispatcher.py [hunk 11]
```diff
@@ -101,6 +103,18 @@ def get_patch_map():
...
+            [
+                (
+                    covariance_module,
+                    "EmpiricalCovariance",
+                    EmpiricalCovariance_sklearnex,
+                ),
+                None,
+            ]
+        ]
+
```
Comment (ethanglaser) [inline sklearnex/dispatcher.py:117]: mapping["empiricalcovariance"] = ...

**Review by icfaust — CHANGES_REQUESTED** [review]
Please include
"covariance/tests": "test_covariance.py",

in this tests_map dict: 
https://github.com/olegkkruglov/scikit-learn-intelex/blob/cov-batch-to-algs/.ci/scripts/select_sklearn_tests.py#L39

and re-run CI's

**Review by icfaust — DISMISSED** [review]

**Review by icfaust — COMMENTED** [review]
accidental approval, my apologies

File: sklearnex/covariance/covariance.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                        self.store_precision == False,
+                        "precision matrix calculation is not supported on oneDAL side",
+                    ),
+                    (not sp.issparse(X), "X is sparse. Sparse input is not supported."),
+                ]
+            )
+            return patching_status
+        raise RuntimeError(f"Unknown method {method_name} in {self.__class__.__name__}")
+
+    def _onedal_cpu_supported(self, method_name, *data):
```
Comment (icfaust) [inline sklearnex/covariance/covariance.py:74]: _onedal_cpu_supported = _onedal_supported
_onedal_gpu_supported = _onedal_supported

File: sklearnex/covariance/covariance.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===============================================================================
+
+from scipy import sparse as sp
+from sklearn.covariance import EmpiricalCovariance as sklearn_EmpiricalCovariance
+
```
Comment (icfaust) [inline sklearnex/covariance/covariance.py:20]:
```python
from sklearn.utils import check_array
```

File: sklearnex/covariance/covariance.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    def _onedal_covariance(self, **onedal_params):
+        return onedal_EmpiricalCovariance(**onedal_params)
+
+    def _onedal_fit(self, X, queue=None):
+        onedal_params = {
+            "method": "dense",
+            "bias": True,
+        }
+
```
Comment (icfaust) [inline sklearnex/covariance/covariance.py:46]:
```python
if sklearn_check_version("0.23"):
   self._validate_data(X)
else:
   check_array(X)
```

**Review by icfaust — CHANGES_REQUESTED** [review]

File: sklearnex/covariance/covariance.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+from .._utils import PatchingConditionsChain
+
+
+class EmpiricalCovariance(sklearn_EmpiricalCovariance):
+    def __init__(self, *, store_precision=True, assume_centered=False):
+        self.store_precision = store_precision
+        self.assume_centered = assume_centered
+
+    def _save_attributes(self):
+        assert hasattr(self, "_onedal_estimator")
```
[the remark this answers is not in the document]
Comment (olegkkruglov) [reply sklearnex/covariance/covariance.py:34]: it is not the only place where it is done like that. if we don't want such style i think it is better to open different pr with fixing it in the whole repo.

**Comment by samir-nasibli** [comment]
@olegkkruglov 
There are currently no benchmark tests in the [sklearn_bench](https://github.com/IntelPython/scikit-learn_bench/) tool for EmpricialCovarience

Maybe it makes sense to leave this function in preview until there are no proofed benchmark numbers?

File: sklearnex/covariance/covariance.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (icfaust) [reply sklearnex/covariance/covariance.py:34]: sklearnex assert calls exist in that aren't tests:
linear.py (linear_regression)
logistic_regression.py (logistic regression preview)
k_means.py (k-means preview)
dbscan.py (dbscan)

This is not the case for all other algorithms in sklearnex (not daal4py). I do think the precedence of this is not consistent, and there is a benefit to our users.

[the remark this answers is not in the document]
Comment (olegkkruglov) [reply sklearnex/covariance/covariance.py]: ok, I will create a task to fix it

File: daal4py/sklearn/monkeypatch/dispatcher.py [hunk 16]
```diff
@@ -161,6 +161,7 @@ def enable(name=None, verbose=True, deprecation=True, get_map=_get_map_of_algori
     if name is not None:
         do_patch(name, get_map)
     else:
+        map_ = get_map()
```
Comment (samir-nasibli) [inline daal4py/sklearn/monkeypatch/dispatcher.py:164]: please remove

File: onedal/covariance/covariance.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        X = make2d(X)
+        types = [np.float32, np.float64]
+        if get_dtype(X) not in types:
+            X = X.astype(np.float64)
+        X = _convert_to_supported(policy, X)
+        dtype = get_dtype(X)
+        params = self._get_onedal_params(dtype)
+        module = _backend.covariance
+        table_X = to_table(X)
+        result = module.compute(policy, params, table_X)
```
Comment (Alexsandruss) [inline onedal/covariance/covariance.py:96]: Hyperparameters setting is required.

File: sklearnex/preview/covariance/covariance.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+from sklearn.utils import check_array
+
+from daal4py.sklearn._utils import sklearn_check_version
+from onedal.covariance import EmpiricalCovariance as onedal_EmpiricalCovariance
+
+from ..._device_offload import dispatch, wrap_output_data
+from ..._utils import PatchingConditionsChain
+
+
+class EmpiricalCovariance(sklearn_EmpiricalCovariance):
```
Comment (Alexsandruss) [inline sklearnex/preview/covariance/covariance.py:29]: Use n_jobs and hyperparameters decorators. For example, like in PCA: https://github.com/intel/scikit-learn-intelex/blob/5b308bec7ea7e65223d026c41370f1ae5ee05db1/sklearnex/preview/decomposition/pca.py#L46-L47

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

**Comment by napetrov** [comment]
@Alexsandruss  @icfaust - i've dismissed review, but be free to provide new round of comments.

File: sklearnex/preview/covariance/covariance.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+from daal4py.sklearn._utils import control_n_jobs, run_with_n_jobs, sklearn_check_version
+from onedal.common.hyperparameters import get_hyperparameters
+from onedal.covariance import EmpiricalCovariance as onedal_EmpiricalCovariance
+
+from ..._device_offload import dispatch
+from ..._utils import PatchingConditionsChain, register_hyperparameters
+
+
+@register_hyperparameters({"fit": get_hyperparameters("covariance", "compute")})
+@control_n_jobs
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/covariance.py:31]: How we can test it?

Comment (samir-nasibli) [reply sklearnex/preview/covariance/covariance.py:31]: I don't get this changes. I would recommend to revert them and if it is required add it in seperate PR.
It is need good description and test cov for this as well.

File: onedal/covariance/tests/test_covariance.py [hunk 20]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
```
Comment (samir-nasibli) [inline onedal/covariance/tests/test_covariance.py:1]:
```suggestion
```

File: sklearnex/preview/covariance/__init__.py [hunk 21]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/__init__.py:1]:
```suggestion
```

File: sklearnex/preview/covariance/covariance.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/covariance.py:1]:
```suggestion
```

File: sklearnex/preview/covariance/covariance.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+from daal4py.sklearn._utils import sklearn_check_version
+from onedal.common.hyperparameters import get_hyperparameters
+from onedal.covariance import EmpiricalCovariance as onedal_EmpiricalCovariance
+
+from ..._device_offload import dispatch
+from ..._utils import PatchingConditionsChain, register_hyperparameters
+
+
+@register_hyperparameters({"fit": get_hyperparameters("covariance", "compute")})
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/covariance.py:30]: Remove also related decorators.

File: sklearnex/preview/covariance/covariance.py [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+from scipy import sparse as sp
+from sklearn.covariance import EmpiricalCovariance as sklearn_EmpiricalCovariance
+from sklearn.utils import check_array
+
+from daal4py.sklearn._utils import sklearn_check_version
+from onedal.common.hyperparameters import get_hyperparameters
+from onedal.covariance import EmpiricalCovariance as onedal_EmpiricalCovariance
+
+from ..._device_offload import dispatch
+from ..._utils import PatchingConditionsChain, register_hyperparameters
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/covariance.py:27]:
```suggestion
from ..._utils import PatchingConditionsChain
```

File: sklearnex/preview/covariance/tests/test_covariance.py [hunk 25]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/tests/test_covariance.py:1]:
```suggestion
```

File: sklearnex/preview/covariance/covariance.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===============================================================================
+
+from scipy import sparse as sp
+from sklearn.covariance import EmpiricalCovariance as sklearn_EmpiricalCovariance
+from sklearn.utils import check_array
+
+from daal4py.sklearn._utils import sklearn_check_version
+from onedal.common.hyperparameters import get_hyperparameters
```
Comment (samir-nasibli) [inline sklearnex/preview/covariance/covariance.py:23]:
```suggestion
```

**Review by samir-nasibli — COMMENTED** [review]
I recommend to move hyper params and n_jobs related changes into separate PRs.

File: sklearnex/preview/covariance/covariance.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+from onedal.covariance import EmpiricalCovariance as onedal_EmpiricalCovariance
+
+from ..._device_offload import dispatch
+from ..._utils import PatchingConditionsChain, register_hyperparameters
+
+
+@register_hyperparameters({"fit": get_hyperparameters("covariance", "compute")})
+class EmpiricalCovariance(sklearn_EmpiricalCovariance):
```
Comment (Alexsandruss) [inline sklearnex/preview/covariance/covariance.py:31]: All patched estimators are required to have `n_jobs` control and hyperparameters (if applicable) decorators

File: sklearnex/preview/covariance/covariance.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+    __doc__ = sklearn_EmpiricalCovariance.__doc__
+
+    def _save_attributes(self):
+        assert hasattr(self, "_onedal_estimator")
+        self.covariance_ = self._onedal_estimator.covariance_
+        self.location_ = self._onedal_estimator.location_
+
+    _onedal_covariance = staticmethod(onedal_EmpiricalCovariance)
+
+    def _onedal_fit(self, X, queue=None):
```
Comment (Alexsandruss) [inline sklearnex/preview/covariance/covariance.py:41]:
```suggestion
    @run_with_n_jobs
    def _onedal_fit(self, X, queue=None):
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: sklearnex/preview/covariance/covariance.py (hunk 27 above)
Comment (samir-nasibli) [reply sklearnex/preview/covariance/covariance.py:31]: Are n_jobs required for Covariance?

Comment (Alexsandruss) [reply sklearnex/preview/covariance/covariance.py:31]: https://github.com/intel/scikit-learn-intelex/blob/61f27db25b62d74a53fb46ec040121670bd3d21f/examples/sklearnex/n_jobs.py#L17-L18

https://github.com/intel/scikit-learn-intelex/blob/61f27db25b62d74a53fb46ec040121670bd3d21f/examples/sklearnex/n_jobs.py#L42-L43

https://github.com/intel/scikit-learn-intelex/blob/61f27db25b62d74a53fb46ec040121670bd3d21f/examples/sklearnex/n_jobs.py#L55-L56

**Comment by samir-nasibli** [comment]
@olegkkruglov
Generally looks good to me.
Hyper params support we can leave in this PR, just update tests.
n_jobs should be moved into seperate PR. I want separate discussion for this.
All changes in should be covered and validated.

File: sklearnex/preview/covariance/covariance.py (hunk 27 above)
Comment (samir-nasibli) [reply sklearnex/preview/covariance/covariance.py:30]: Thnx for links.
https://github.com/intel/scikit-learn-intelex/pull/1480/files#r1426643617
We are failing some tests in this case. Are we going to fix it?

File: sklearnex/preview/covariance/covariance.py (hunk 26 above)
Comment (samir-nasibli) [reply sklearnex/preview/covariance/covariance.py:22]: outdated.

File: sklearnex/preview/covariance/covariance.py (hunk 23 above)
Comment (samir-nasibli) [reply sklearnex/preview/covariance/covariance.py:29]: outdated.

**Review by samir-nasibli — APPROVED** [review]

File: sklearnex/dispatcher.py [hunk 29]
```diff
@@ -73,9 +74,12 @@ def get_patch_map():
         from .neighbors import KNeighborsRegressor as KNeighborsRegressor_sklearnex
         from .neighbors import LocalOutlierFactor as LocalOutlierFactor_sklearnex
         from .neighbors import NearestNeighbors as NearestNeighbors_sklearnex
+        from .preview.cluster import KMeans as KMeans_sklearnex
```
Comment (ethanglaser) [inline sklearnex/dispatcher.py:77]: why is it moved?

File: sklearnex/preview/covariance/tests/test_covariance.py [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+    from sklearnex.preview.covariance import EmpiricalCovariance
+
+    X = np.array([[0, 1], [0, 1]])
+    X = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+    empcov = EmpiricalCovariance()
+    if daal_check_version((2024, "P", 0)) and macro_block is not None:
+        hparams = empcov.get_hyperparameters("fit")
+        hparams.cpu_macro_block = macro_block
+    result = empcov.fit(X)
+    expected_covariance = np.array([[0, 0], [0, 0]])
```
Comment (ethanglaser) [inline sklearnex/preview/covariance/tests/test_covariance.py:40]: Seems like a very basic test - would it make sense to have a bit more of a testing scope? i.e. one whose result isn't just all zeroes

File: sklearnex/dispatcher.py (hunk 29 above)
Comment (olegkkruglov) [reply sklearnex/dispatcher.py:77]: i guess it might be result of applying isort or black. i can't remember i did it. i'll fix it, thanks for mentioning

**Review by ethanglaser — APPROVED** [review]
CI looks good and updates look good to me. If there are remaining comments to be addressed please fix before merging

File: sklearnex/preview/covariance/covariance.py [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+from daal4py.sklearn._utils import sklearn_check_version
+from onedal.common.hyperparameters import get_hyperparameters
+from onedal.covariance import EmpiricalCovariance as onedal_EmpiricalCovariance
+
+from ..._device_offload import dispatch
+from ..._utils import PatchingConditionsChain, register_hyperparameters
+
+
+@register_hyperparameters({"fit": get_hyperparameters("covariance", "compute")})
+class EmpiricalCovariance(sklearn_EmpiricalCovariance):
```
Comment (Alexsandruss) [inline sklearnex/preview/covariance/covariance.py:30]: Comment about `n_jobs` support is still valid: if one estimator is excluded, then extension will lose consistency.

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: sklearnex/preview/covariance/covariance.py (hunk 31 above)
Comment (samir-nasibli) [reply sklearnex/preview/covariance/covariance.py:30]: This request of changes is not critical.
We are agile. We are trying incrementally enable features in our library.
Support of n_jobs added after the estimators support.
Let's first have this pretty good PR merge, and then add n_jobs support in the next PR.

**Review by napetrov — APPROVED** [review]

**Pull request merged**